packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -1440,46 +1444,102 @@ class BuildOwner {
       }
       return true;
     });
-    if (_dirtyElements.isEmpty && onBuildScheduled != null)
+    if (!_scheduledFlushDirtyElements && onBuildScheduled != null) {
+      _scheduledFlushDirtyElements = true;
       onBuildScheduled();
+    }
     _dirtyElements.add(element);
+    element._inDirtyList = true;
   }
 
   int _debugStateLockLevel = 0;
   bool get _debugStateLocked => _debugStateLockLevel > 0;
   bool _debugBuilding = false;
   BuildableElement _debugCurrentBuildTarget;
 
-  /// Establishes a scope in which calls to [State.setState] are forbidden.
+  /// Establishes a scope in which calls to [State.setState] are forbidden, and
+  /// calls the given `callback`.
+  ///
+  /// This mechanism is used to ensure that, for instance, [State.dispose] does
+  /// not call [State.setState].
+  void lockState(void callback()) {
+    assert(callback != null);
+    assert(_debugStateLockLevel >= 0);
+    assert(() {
+      _debugStateLockLevel += 1;
+      return true;
+    });
+    try {
+      callback();
+    } finally {
+      assert(() {
+        _debugStateLockLevel -= 1;
+        return true;
+      });
+    }
+    assert(_debugStateLockLevel >= 0);
+  }
+
+  /// Establishes a scope for updating the widget tree, and calls the given
+  /// `callback`, if any. Then, builds all the elements that were marked as
+  /// dirty using [scheduleBuildFor], in depth order.
   ///
   /// This mechanism prevents build functions from transitively requiring other
   /// build functions to run, potentially causing infinite loops.
   ///
-  /// If the building argument is true, then this function enables additional
-  /// asserts that check invariants that should apply during building.
+  /// The dirty list is processed after `callback` returns, building all the
+  /// elements that were marked as dirty using [scheduleBuildFor], in depth
+  /// order. If elements are marked as dirty while this method is running, they
+  /// must be deeper than the `context` node, and deeper than any
+  /// previously-built node in this pass.
+  ///
+  /// To flush the current dirty list without performing any other work, this
+  /// function can be called with no callback. This is what the framework does
+  /// each frame, in [WidgetsBinding.beginFrame].
   ///
-  /// The context argument is used to describe the scope in case an exception is
-  /// caught while invoking the callback.
-  void lockState(void callback(), { bool building: false }) {
-    bool debugPreviouslyBuilding;
+  /// Only one [buildScope] can be active at a time.
+  ///
+  /// A [buildScope] implies a [lockState] scope as well.
+  void buildScope(Element context, [VoidCallback callback]) {
```

Comment:
This seems like a bit of a misnomer, since buildScope() doesn't just establish a scope for its callback.